### PR TITLE
Create a media selector #193

### PR DESCRIPTION
**Pull request by hbaptiste**
New media Library for the core that allows us to create and edit a media item according to its type.

### CODE REVIEW

**Comment by mickaelandrieu** [comment]
@hbaptiste your tests seems broken.

File: src/tb/component/medialibrary/helpers/contextmenu.helper.js [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+                btnLabel: "Edit",
+                btnCallback: actions.editAction
+            });
+
+            mediaFolderContextMenu.addMenuItem({
+                btnCls: "bb5-context-menu-remove",
+                btnLabel: "Remove",
+                btnCallback: actions.removeAction
+            });
+            /*  mediaFolderContextMenu.addMenuItem({
```
Comment (mickaelandrieu) [inline src/tb/component/medialibrary/helpers/contextmenu.helper.js]: ?

**Comment by mickaelandrieu** [comment]
Also, this need tests.

File: src/tb/component/medialibrary/helpers/contextmenu.helper.js (hunk 1 above)
Comment (ndufreche) [inline src/tb/component/medialibrary/helpers/contextmenu.helper.js]: remove commented code please.

File: src/tb/component/medialibrary/helpers/contextmenu.helper.js [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+            selectedNode = node;
+        },
+        getContextMenu: function () {
+            if (!contextMenu) {
+                contextMenu = buildContextMenu();
+            }
+            return contextMenu;
+        }
+    };
+});
```
Comment (ndufreche) [inline src/tb/component/medialibrary/helpers/contextmenu.helper.js]: Add some spaces in this file please.

File: src/tb/component/medialibrary/main.js [hunk 3]
```diff
@@ -0,0 +1,396 @@
...
+        },
+        resetOnClose: true,
+        categoryTreeView: {}
+    },
+        MediaLibrary = new JS.Class({
+            initialize: function (config) {
+                jQuery.extend(this, {}, BackBone.Events);
+                this.config = config || {};
+                this.VIEW_MODE = 'view';
+                this.EDIT_MODE = 'edit';
```
Comment (ndufreche) [inline src/tb/component/medialibrary/main.js]: If this is a constant remove it from the initialize please

File: src/tb/component/medialibrary/main.js [hunk 4]
```diff
@@ -0,0 +1,396 @@
...
+            },
+            getAvailableMedia: function () {
+                return this.config.available_media;
+            },
+            selectItems: function (media) {
+                this.mediaListView.selectItems(media);
+            },
+            addButtons: function () {
+                this.dialog.addButton("Yes", function () {
+                    alert("alert Yes");
```
Comment (ndufreche) [inline src/tb/component/medialibrary/main.js]: alert no no no no

File: src/tb/component/medialibrary/main.js [hunk 5]
```diff
@@ -0,0 +1,396 @@
...
+                });
+                this.dialog.addButton("No", function () {
+                    alert("alert Yes");
```
Comment (ndufreche) [inline src/tb/component/medialibrary/main.js]: alert no no no no

File: src/tb/component/medialibrary/main.js [hunk 6]
```diff
@@ -0,0 +1,396 @@
...
+                                deps.EditionHelper.show(content, {
+                                    onSave: jQuery.proxy(self.onSaveHandler, self, mediaInfos)
+                                });
+                            }).fail(function (reason) {
+                                require("component!notify").error(reason);
+                            });
+                        }
+                    });
+                } catch (e) {
+                    Api.exception("MediaLibrary:Exception", 64535, e);
```
Comment (ndufreche) [inline src/tb/component/medialibrary/main.js]: Add a error message and pass the e message into the parameters.
Api.exception("MediaLibraryException", 64535, "Error was raise during the generation of the media form", {error: e});

File: src/tb/component/medialibrary/main.js [hunk 7]
```diff
@@ -0,0 +1,396 @@
...
+                    parentNode = this.openedMediaFolder || this.selectedMediaFolder;
+                }
+                this.mediaFolderTreeView.setData(formattedData, parentNode);
+                if (parentNode) {
+                    parentNode.isLoaded = true;
+                }
+                /*if (parentNode) {
+                    this.loadingMap[parentNode.uid] = parentNode.uid;
+                    parentNode.isLoaded = true;
+                }*/
```
Comment (ndufreche) [inline src/tb/component/medialibrary/main.js]: Remove commented code.

File: src/tb/component/medialibrary/main.js [hunk 8]
```diff
@@ -0,0 +1,396 @@
...
+            },
+            handleMediaSelection: function (e) {
+                /*if (jQuery(e.node.element).hasClass("jqtree-selected")) {
+                    return false;
+                }*/
```
Comment (ndufreche) [inline src/tb/component/medialibrary/main.js]: Remove commented code.

File: src/tb/component/medialibrary/main.js [hunk 9]
```diff
@@ -0,0 +1,396 @@
...
+        },
+        createMediaLibrary: function (userConfig) {
+            userConfig = userConfig || {};
+            var config = jQuery.extend(true, defaultConfig, userConfig),
+                mediaLibrary = new MediaLibrary(config);
+            return mediaLibrary;
+        },
+        MediaLibrary: MediaLibrary
+    };
+});
```
Comment (ndufreche) [inline src/tb/component/medialibrary/main.js]: Add some spaces into that file please.

File: src/tb/component/pagination/main.js [hunk 10]
```diff
@@ -147,43 +145,34 @@ define(['jquery', 'jssimplepagination'], function (jQuery) {
...
             enable: function () {
                 this.invoke('enable');
             },
-
             getItemsOnPage: function () {
                 var conf = this.getPaginationConf();
-
                 return conf.itemsOnPage;
             }
         });
```
Comment (ndufreche) [inline src/tb/component/pagination/main.js]: Re add spaces pleases.

File: src/tb/component/searchengine/main.js [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+        }
+    });
+    return {
+        createSimpleSearchEngine: function (config) {
+            config = config || {};
+            return new SimpleSearchEngine(config);
+        },
+        SimpleSearchEngine: SimpleSearchEngine
+    };
+});
```
Comment (ndufreche) [inline src/tb/component/searchengine/main.js]: Add some spaces in this file please.

File: src/tb/component/treeview/NodeEditor.js [hunk 12]
```diff
@@ -0,0 +1,93 @@
...
+            if (!this.tree.isNodeOpened(selectedNode)) {
+                this.tree.invoke('openNode', selectedNode);
+            }
+
+            node = this.tree.getNodeById('node-editor');
+            this.edit(node);
+        },
+
+        bindEvents: function () {
+            jQuery(this.tree.el).on('click', '.fa-check', jQuery.proxy(this.handleEdition, this));
```
Comment (ndufreche) [inline src/tb/component/treeview/NodeEditor.js]: Use specific class not a style element.

File: src/tb/component/treeview/NodeEditor.js [hunk 13]
```diff
@@ -0,0 +1,93 @@
...
+            jQuery(this.tree.el).on('click', '.fa-close', jQuery.proxy(this.cancelEdition, this));
```
Comment (ndufreche) [inline src/tb/component/treeview/NodeEditor.js]: Use specific class not a style element and if is possible prefer ids

File: src/tb/core/RestDriver.js [hunk 14]
```diff
@@ -120,7 +120,6 @@ define('tb.core.RestDriver', ['tb.core.Request', 'tb.core.RequestHandler', 'URIj
                 if (null === criterias) {
                     return this;
                 }
-
```
Comment (ndufreche) [inline src/tb/core/RestDriver.js]: Remove this form your PR please

File: src/tb/apps/page/views/page.view.tree.contribution.js [hunk 15]
```diff
@@ -126,6 +126,8 @@ define(
                     parent_uid = moveInfo.moved_node.parent.id,
                     data = {};
 
+
+
```
Comment (ndufreche) [inline src/tb/apps/page/views/page.view.tree.contribution.js]: remove this from your PR please

File: src/tb/component/contextmenu/main.js [hunk 16]
```diff
@@ -97,7 +62,17 @@ define(['jquery', 'jsclass'], function (jQuery) {
          * Bind events
          */
         bindEvents: function () {
+            var self = this;
             jQuery(this.settings.contentSelector).on("contextmenu", jQuery.proxy(this.show, this));
+            jQuery('body').on('contextmenu', function () {
```
Comment (ndufreche) [inline src/tb/component/contextmenu/main.js]: Assign jQuery('body') to a variable.

File: src/tb/core/RestDriver.js (hunk 14 above)
Comment (mickaelandrieu) [inline src/tb/core/RestDriver.js]: a criterion, many criteria (remove 's')

File: src/tb/component/treeview/NodeEditor.js (hunk 13 above)
Comment (hbaptiste) [inline src/tb/component/treeview/NodeEditor.js]: i can't use an id as we can have many instances of the NodeEditor.

File: src/tb/core/RestDriver.js (hunk 14 above)
Comment (hbaptiste) [inline src/tb/core/RestDriver.js]: not my typo!

Comment (ndufreche) [inline src/tb/core/RestDriver.js]: I create an issue for that.

**Comment by ndufreche** [comment]
great job :+1: but i need you rebase it now (it's because you don't remove src/tb/core/RestDriver.js :stuck_out_tongue:)

**Comment by ndufreche** [comment]
@hbaptiste merged thanks for this big contribution.

**Pull request merged**